### PR TITLE
[MDETR] Add contrastive alignment loss

### PR DESCRIPTION
**Pull request by ebsmothers**
Summary:
This PR adds support for the contrastive alignment loss from MDETR. This loss also necessitates
adding two projections to the MDETR model, and returning some intermediate outputs. I modified the
return types of MDETRTransformer and MDETR to be NamedTuples in the process of making these changes.

Note that the contrastive alignment loss uses the mapping between character positions and token IDs
and so needs access to the transformers tokenizer. For this reason I am putting this loss in examples/
instead of in core.

Test plan:
Added a new test for the contrastive loss, updated the existing MDETR model tests
to account for different random initialization orders and new params.
Confirmed in LoadAndComparePretrainedWeights.ipynb that the outputs line up with the MDETR checkpoint.

```
python -m pytest -v
...
========================= 175 passed, 2 skipped, 2 xfailed, 6 warnings in 46.57s ================
```

### CODE REVIEW

**Review by langong347 — COMMENTED** [review]
Would be possible to add comments about dimensions for at least the inputs and the outputs of each class? Image (and video) data are high dimensional and often we rearrange the dimensions within a module. One can quickly lose track of what's going on.

File: examples/mdetr/contrastive_alignment_loss.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    tokens_to_boxes_loss = (
+        ((pos_term / nb_pos + neg_term)).masked_fill(~tokens_with_pos, 0).sum()
+    )
+
+    tot_loss = (box_to_token_loss + tokens_to_boxes_loss) / 2
+
+    return tot_loss / num_boxes
+
+
+def construct_positive_map(logits: Tensor, target_tokens, indices, tokenized):
```
Comment (RdoubleA) [inline examples/mdetr/contrastive_alignment_loss.py:84]: type annotations and return type

File: examples/mdetr/contrastive_alignment_loss.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+def contrastive_alignment_loss(
+    projected_queries: Tensor,
+    projected_tokens: Tensor,
+    target_tokens: List[List[List[int]]],
+    indices: List[Tuple[Tensor, Tensor]],
+    num_boxes: int,
+    tokenized: BatchEncoding,
+    temperature: float = 0.07,
+):
```
Comment (RdoubleA) [inline examples/mdetr/contrastive_alignment_loss.py:22]: return type

File: examples/mdetr/contrastive_alignment_loss.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        for j, tok_list in enumerate(cur_tokens):
+            for (beg, end) in tok_list:
+                beg_pos = tokenized.char_to_token(i, beg)
+                end_pos = tokenized.char_to_token(i, end - 1)
+                if beg_pos is None:
+                    try:
+                        beg_pos = tokenized.char_to_token(beg + 1)
+                        if beg_pos is None:
+                            beg_pos = tokenized.char_to_token(beg + 2)
+                    except Exception:
```
Comment (RdoubleA) [inline examples/mdetr/contrastive_alignment_loss.py:99]: why are we checking offsets of beg and end? is it to find multiple tokens that are associated with the query? then why only upto offset of 3?

File: examples/mdetr/contrastive_alignment_loss.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+
+def construct_positive_map(logits: Tensor, target_tokens, indices, tokenized):
+    # construct a map such that positive_map[k, i,j] = True iff query i is associated to token j in batch item k
+    # For efficency, the construction happens on CPU, then the whole matrix is transferred to GPU in one go.
+    positive_map = torch.zeros(logits.shape, dtype=torch.bool)
+    for i, ((idx_src, idx_tgt), tgt) in enumerate(zip(indices, target_tokens)):
+        cur_tokens = [tgt[j] for j in idx_tgt]
+        for j, tok_list in enumerate(cur_tokens):
+            for (beg, end) in tok_list:
```
[the remark this answers is not in the document]
Comment (ebsmothers) [reply examples/mdetr/contrastive_alignment_loss.py:91]: So the outer loop is for each sample in the batch. The inner two are both related to how the tokens are associated to boxes. In general a given box may be represented by multiple phrases in the caption, which is why they are represented as a set of disjoint intervals in the token space. So the middle loop iterates over boxes, while the innermost loop iterates over these disjoint intervals. In practice, it is pretty rare that a given box has multiple intervals like this, but it does happen.

TLDR: I don't think we can condense it 😅, but it is most often a no-op anyways

File: examples/mdetr/contrastive_alignment_loss.py (hunk 3 above)
Comment (ebsmothers) [reply examples/mdetr/contrastive_alignment_loss.py:99]: I think this is just some heuristic they use for error handling. Actually I just tested on the phrase grounding eval code and it never even hits this portion of the code (the first char_to_token call always works). So I will take it out to keep things simpler, if we find that we need it for VQA or something we can always add it back later on

File: examples/mdetr/contrastive_alignment_loss.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    tokens_to_boxes_loss = (
+        ((pos_term / nb_pos + neg_term)).masked_fill(~tokens_with_pos, 0).sum()
+    )
+
+    tot_loss = (box_to_token_loss + tokens_to_boxes_loss) / 2
+
+    return tot_loss / num_boxes
+
+
+def construct_positive_map(
```
Comment (RdoubleA) [inline examples/mdetr/contrastive_alignment_loss.py:84]: add a test for this?

File: examples/mdetr/test/test_contrastive_alignment_loss.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+# LICENSE file in the root directory of this source tree.
+
+import pytest
+import torch
+from examples.mdetr.contrastive_alignment_loss import contrastive_alignment_loss
+from test.test_utils import assert_expected, set_rng_seed
+from transformers import RobertaTokenizerFast
+
+
+@pytest.fixture(scope="class", autouse=True)
```
Comment (RdoubleA) [inline examples/mdetr/test/test_contrastive_alignment_loss.py:14]: we decided on using function scope as default right? but if this changes all the output values then let's leave it

**Review by RdoubleA — APPROVED** [review]

File: examples/mdetr/contrastive_alignment_loss.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    positive_logits = -logits.masked_fill(~positive_map, 0)
+    negative_logits = logits
+
+    # Calculate the contrastive loss for all objects
+    boxes_with_pos = positive_map.any(2)
+    pos_term = positive_logits.sum(2)
+    neg_term = negative_logits.logsumexp(2)
+    nb_pos = positive_map.sum(2) + 1e-6
+    box_to_token_loss = (
+        ((pos_term / nb_pos + neg_term)).masked_fill(~boxes_with_pos, 0).sum()
```
Comment (ankitade) [inline examples/mdetr/contrastive_alignment_loss.py:67]: why does nb_pos not divide neg_term also?

**Review by ankitade — APPROVED** [review]
